### PR TITLE
Support prroi_pool_op with Tensor and LoDTensor rois

### PR DESCRIPTION
**Pull request by LDOUBLEV**
1. Add a new input named batch_roi_nums for prroi_pool_op. batch_roi_nums includes the number of roi for each image in batch when rois is Tensor.  This information is saved in rois's lod when rois is LoDTensor.

2. add grad check to prroi_pool_op and solve unnormal X grad diff in CPU.

image

image

### CODE REVIEW

File: paddle/fluid/operators/prroi_pool_op.cc [hunk 1]
```diff
@@ -39,6 +39,11 @@ class PRROIPoolOpMaker : public framework::OpProtoAndCheckerMaker {
              "where (x1, y1) is the top left coordinates, and "
              "(x2, y2) is the bottom right coordinates. "
              "The roi batch index can be calculated from LoD.");
+    AddInput("Batch_index",
```
Comment (qingqing01) [inline paddle/fluid/operators/prroi_pool_op.cc]: Batch_index  -> BatchIndex

觉用BatchRoINums，也就是长度，和LoD对应，而不是BatchIndex。 这样和其他seq的padding操作行为一致，更合理些~

Comment (LDOUBLEV) [reply paddle/fluid/operators/prroi_pool_op.cc]: OK

File: paddle/fluid/operators/prroi_pool_op.cc [hunk 2]
```diff
@@ -108,6 +112,13 @@ class PRROIPoolOp : public framework::OperatorWithKernel {
     out_dims[1] = input_dims[1];
     out_dims[2] = pooled_height;
     out_dims[3] = pooled_width;
+    bool has_batch_index = ctx->HasInput("BatchRoINums");
+    if (has_batch_index) {
```
Comment (heavengate) [inline paddle/fluid/operators/prroi_pool_op.cc]: 建议 if (ctx->HasInput("BatchRoINums")) 代码尽量简洁易读~

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py [hunk 3]
```diff
@@ -82,6 +82,7 @@ def test_check_output(self):
         self.check_output()
 
     def test_backward(self):
+        #self.check_grad(['X'], 'Out')
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: remove

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py [hunk 4]
```diff
@@ -130,5 +131,123 @@ def test_errors(self):
...
+            rois_index = fluid.layers.data(
+                name='rois_idx', shape=[], dtype="int64")
+            output = fluid.layers.prroi_pool(
+                x, rois, 0.25, 2, 2, batch_roi_nums=rois_index)
+            loss = fluid.layers.mean(output)
+            optimizer = fluid.optimizer.SGD(learning_rate=1e-3)
+            optimizer.minimize(loss)
+            #input_x = fluid.create_lod_tensor(self.x, [], place)
+            #input_rois = fluid.create_lod_tensor(self.rois[:, 1:5],
+            #                                     [], place)
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: remove comments

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -16256,26 +16257,42 @@ def prroi_pool(input,
...
-
+            ## example one
             import paddle.fluid as fluid
-            x = fluid.layers.data(name='x', shape=[490, 28, 28], dtype='float32')
-            rois = fluid.layers.data(name='rois', shape=[4], lod_level=1, dtype='float32')
+            x = fluid.data(name='x', shape=[None, 490, 28, 28], dtype='float32')
+            rois = fluid.data(name='rois', shape=[None, 4], lod_level=1, dtype='float32')
             pool_out = fluid.layers.prroi_pool(x, rois, 1.0, 7, 7)
+            
+            ## example two
```
Comment (heavengate) [inline python/paddle/fluid/layers/nn.py]: 建议上面example one删掉，这里example two改成prroi_pool with batch_roi_num 之类的

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py (hunk 3 above)
Comment (heavengate) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: 这个OP没有梯度单测？

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py [hunk 6]
```diff
@@ -82,6 +82,7 @@ def test_check_output(self):
         self.check_output()
 
     def test_backward(self):
+        #self.check_grad(['X'], 'Out')
         for place in self._get_places():
             self._get_gradient(['X'], place, ["Out"], None)
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: 这里没有检查解析梯度数值正确性？

Comment (LDOUBLEV) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: 这里有点奇怪，使用self.check_grad(['X'], 'Out')，单测不通过，但是我本地测试梯度结果能和pytorch版本的结果对齐，self._get_gradient(['X'], place, ["Out"], None)单测可以过

File: paddle/fluid/operators/prroi_pool_op.cc (hunk 2 above)
Comment (LDOUBLEV) [reply paddle/fluid/operators/prroi_pool_op.cc]: Done!

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py (hunk 3 above)
Comment (LDOUBLEV) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: Done

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py (hunk 4 above)
Comment (LDOUBLEV) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 5 above)
Comment (LDOUBLEV) [reply python/paddle/fluid/layers/nn.py]: example one还是保留了，改成了prroi_pool without batch_roi_num,    example two改成了prroi_pool with batch_roi_num

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py (hunk 6 above)
Comment (heavengate) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: `_get_gradient`只是调用了反向计算了梯度，并没有校验梯度计算结果是否正确，可以排查一下为啥单测会不通过，或者通过改大atol误差阈值这些方法调通单测，最好还是有单测，以免后面有其他人的修改其他模块影响到你这个OP之类的

Comment (LDOUBLEV) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: Thanks, 梯度变化较为剧烈，delta等于0.005的情况下max gradient diff 0.999983，目前设置了check_grad() max_relative_error为1.0，单测可以过

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py [hunk 7]
```diff
@@ -130,5 +129,119 @@ def test_errors(self):
...
+
+    def setUp(self):
+        self.op_type = 'prroi_pool'
+        self.set_data()
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_backward(self):
+        self.check_grad(['X'], 'Out', max_relative_error=1.0)
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: max_relative_error有点大，分析下原因，可以尝试double输入类型看下

File: paddle/fluid/operators/prroi_pool_op.cu [hunk 8]
```diff
@@ -214,23 +214,42 @@ class GPUPRROIPoolOpKernel : public framework::OpKernel<T> {
...
+          rois_batch_size, batch_size,
+          "The rois_batch_size and input(X) batch_size must be the same.");
+      int rois_num_with_lod = rois_lod[rois_batch_size];
+      PADDLE_ENFORCE_EQ(rois_num, rois_num_with_lod,
+                        "The rois_num from input and lod must be the same.");
+
+      for (int n = 0; n < rois_batch_size; ++n) {
+        for (size_t i = rois_lod[n]; i < rois_lod[n + 1]; ++i) {
+          rois_batch_id_data[i] = n;
+        }
```
Comment (qingqing01) [inline paddle/fluid/operators/prroi_pool_op.cu:255]: 发现下面 rois_batch_id_list_gpu用法不合理，需参考roi_pooling, roi_algin修改，

    auto roi_ptr = memory::Alloc(dev_ctx, bytes);
    int* roi_id_data = reinterpret_cast<int*>(roi_ptr->ptr());

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py (hunk 7 above)
Comment (LDOUBLEV) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: double输入类型的grad的diff也是很大，我再分析下是不是其他问题导致的

File: paddle/fluid/operators/prroi_pool_op.cu (hunk 8 above)
Comment (LDOUBLEV) [reply paddle/fluid/operators/prroi_pool_op.cu:255]: Done

**Review by qingqing01 — DISMISSED** [review]

**Review by heavengate — DISMISSED** [review]

**Review by wanghaoshuang — DISMISSED** [review]

**Review by qingqing01 — DISMISSED** [review]

File: paddle/fluid/operators/prroi_pool_op.cu [hunk 9]
```diff
@@ -214,41 +214,66 @@ class GPUPRROIPoolOpKernel : public framework::OpKernel<T> {
...
     int* rois_batch_id_data =
         rois_batch_id_list.mutable_data<int>(platform::CPUPlace());
-    for (int n = 0; n < rois_batch_size; ++n) {
-      for (size_t i = rois_lod[n]; i < rois_lod[n + 1]; ++i) {
-        rois_batch_id_data[i] = n;
+
+    if (ctx.HasInput("BatchRoINums") || rois->lod().empty()) {
+      auto* BatchIndex = ctx.Input<Tensor>("BatchRoINums");
+      framework::Tensor batch_index_cpu;
+      framework::TensorCopy(*BatchIndex, platform::CPUPlace(),
```
Comment (qingqing01) [inline paddle/fluid/operators/prroi_pool_op.cu]: Use TensorCopySync here

File: paddle/fluid/operators/prroi_pool_op.cu [hunk 10]
```diff
@@ -214,41 +214,66 @@ class GPUPRROIPoolOpKernel : public framework::OpKernel<T> {
...
     // set rois batch id
     framework::Tensor rois_batch_id_list;
     rois_batch_id_list.Resize({rois_num});
     int* rois_batch_id_data =
         rois_batch_id_list.mutable_data<int>(platform::CPUPlace());
-    for (int n = 0; n < rois_batch_size; ++n) {
-      for (size_t i = rois_lod[n]; i < rois_lod[n + 1]; ++i) {
-        rois_batch_id_data[i] = n;
+
+    if (ctx.HasInput("BatchRoINums") || rois->lod().empty()) {
```
Comment (qingqing01) [inline paddle/fluid/operators/prroi_pool_op.cu]: line 219 行，如果有`BatchRoINums`作为输入的话， `rois`输入是经过padding的3D Tensor吧，如果是上面`rois_num`计算有误， backbward存在同样的问题

File: paddle/fluid/operators/prroi_pool_op.cu [hunk 11]
```diff
@@ -214,41 +214,66 @@ class GPUPRROIPoolOpKernel : public framework::OpKernel<T> {
...
+      auto* BatchIndex = ctx.Input<Tensor>("BatchRoINums");
```
Comment (qingqing01) [inline paddle/fluid/operators/prroi_pool_op.cu]: BatchIndex注意命名规范 https://google.github.io/styleguide/cppguide.html#Variable_Names

lowercase

Comment (LDOUBLEV) [reply paddle/fluid/operators/prroi_pool_op.cu]: Done

File: paddle/fluid/operators/prroi_pool_op.cu [hunk 12]
```diff
@@ -281,17 +306,40 @@ class GPUPRROIPoolGradOpKernel : public framework::OpKernel<T> {
       rois_batch_id_list.Resize({rois_num});
       int* rois_batch_id_data =
           rois_batch_id_list.mutable_data<int>(platform::CPUPlace());
-      auto rois_lod = rois->lod().back();
-      int rois_batch_size = rois_lod.size() - 1;
-      for (int n = 0; n < rois_batch_size; ++n) {
-        for (size_t i = rois_lod[n]; i < rois_lod[n + 1]; ++i) {
-          rois_batch_id_data[i] = n;
+
+      if (ctx.HasInput("BatchRoINums") || rois->lod().empty()) {
```
Comment (liym27) [inline paddle/fluid/operators/prroi_pool_op.cu:310]: Errors should be reported if  `rois->lod().empty() == true` but `ctx.HasInput("BatchRoINums") == false`

Comment (LDOUBLEV) [reply paddle/fluid/operators/prroi_pool_op.cu:310]: Done! thanks

**Review by qingqing01 — DISMISSED** [review]

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py [hunk 13]
```diff
@@ -82,8 +83,16 @@ def test_check_output(self):
         self.check_output()
 
     def test_backward(self):
-        for place in self._get_places():
-            self._get_gradient(['X'], place, ["Out"], None)
+        grad_diff = [0.007, 0.007]
```
Comment (lanxianghit) [inline python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: 精度阈值太大 如何确定计算准确性？

Comment (LDOUBLEV) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: 这个OP和竞品Pytorch的Precise roi pooling的相对误差和绝对误差在1e-5量级，并且在ATOM模型中使用这个OP可以在精度上和竞品打平。

**Comment by LDOUBLEV** [comment]
补充说明：
本次修改中，prroi_pool单测的max_releative_error 设置为0.007，大于0.005。
本次测试结果：实际测试发现，误差在0.005和0.0063之间。
与竞品对比：与竞品pytorch的precise roi pooling实现对比，相对误差和绝对误差均在1e-5量级。竞品实现链接-https://github.com/vacancy/PreciseRoIPooling
模型应用结果：在模型ATOM中使用prroi_pool，可以和pytorch  ATOM的精度持平。ATOM 模型将随paddle 1.7发版

File: paddle/fluid/operators/prroi_pool_op.cc [hunk 14]
```diff
@@ -39,6 +39,11 @@ class PRROIPoolOpMaker : public framework::OpProtoAndCheckerMaker {
              "where (x1, y1) is the top left coordinates, and "
              "(x2, y2) is the bottom right coordinates. "
              "The roi batch index can be calculated from LoD.");
+    AddInput("BatchRoINums",
+             "(Tensor), "
+             "1-D tensor with shape [N], the number of"
+             " rois for each img in batch, where N is the batch size")
```
Comment (Boyan-Liu) [inline paddle/fluid/operators/prroi_pool_op.cc]: rois -> ROIs，请与下面保持一致
img -> image，请使用全称

**Review by Boyan-Liu — DISMISSED** [review]
请1.7后记得修复

File: paddle/fluid/operators/prroi_pool_op.cc (hunk 14 above)
Comment (LDOUBLEV) [reply paddle/fluid/operators/prroi_pool_op.cc]: 收到

**Review by Boyan-Liu — APPROVED** [review]

File: paddle/fluid/operators/prroi_pool_op.cc (hunk 14 above)
Comment (LDOUBLEV) [reply paddle/fluid/operators/prroi_pool_op.cc]: Done

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py (hunk 13 above)
Comment (LDOUBLEV) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: 经过排查，grad diff过大的原因和输入数据的shape有关，输入shape过大过小都存在grad diff过大的问题，shape过小出现diff过大的原因可能是roi的数据范围与小的输入数据shape不匹配，因此在单测中调整了spatial_scale这个参数控制roi的范围，并增大输入shape，可以解决这个问题。并且，与pytorch对比的结果也是正常的。

**Comment by LDOUBLEV** [comment]
> 补充说明：
> 本次修改中，prroi_pool单测的max_releative_error 设置为0.007，大于0.005。
> 本次测试结果：实际测试发现，误差在0.005和0.0063之间。
> 与竞品对比：与竞品pytorch的precise roi pooling实现对比，相对误差和绝对误差均在1e-5量级。竞品实现链接-https://github.com/vacancy/PreciseRoIPooling
> 模型应用结果：在模型ATOM中使用prroi_pool，可以和pytorch ATOM的精度持平。ATOM 模型将随paddle 1.7发版

最新测试结果，gradiff的问题已经解决，误差小于默认值，与竞品能对齐

**Review by lanxianghit — APPROVED** [review]
